Content.Client/Actions/ActionsSystem.cs

```diff
@@ -64,12 +110,172 @@ public override void Initialize()
+        {
+            if (!Resolve(uid, ref comp, false))
+                return;
+
+            base.AddActions(uid, actions, provider, comp, dirty);
+            UIDirty = true;
+        }
+
+        public override void RemoveActions(EntityUid uid, IEnumerable<ActionType> actions, ActionsComponent? comp = null, bool dirty = true)
+        {
+            if (!Resolve(uid, ref comp, false))
+                return;
+
+            base.RemoveActions(uid, actions, comp, dirty);
+            UIDirty = true;
+        }
+
+        public override void FrameUpdate(float frameTime)
+        {
+            // avoid updating GUI when doing predictions & resetting state.
+            if (UIDirty)
+            {
+                UIDirty = false;
+                UpdateUI();
+            }
+        }
+
+        /// <summary>
+        /// Updates the displayed hotbar (and menu) based on current state of actions.
+        /// </summary>
+        public void UpdateUI()
+        {
+            if (Ui == null)
+                return;
+
+            foreach (var action in Ui.Component.Actions)
+            {
+                if (action.AutoPopulate && !Assignments.Assignments.ContainsKey(action))
+                    Assignments.AutoPopulate(action, Ui.SelectedHotbar, false);
+            }
+
+            // get rid of actions that are no longer available to the user
+            foreach (var (action, index) in Assignments.Assignments.ToList())
+            {
+                if (index.Count == 0)
+                {
+                    Assignments.Assignments.Remove(action);
+                    continue;
+                }
+
+                if (Ui.Component.Actions.Contains(action))
+                    continue;
+
+                Assignments.ClearSlot(index[0].Hotbar, index[0].Slot, false);
+            }
+
+            Assignments.PreventAutoPopulate.RemoveWhere(action => !Ui.Component.Actions.Contains(action));
+
+            Ui.UpdateUI();
+        }
+
+        public void HandleHotbarKeybind(byte slot, in PointerInputCmdHandler.PointerInputCmdArgs args)
+        {
+            Ui?.HandleHotbarKeybind(slot, args);
+        }
+
+        public void HandleChangeHotbarKeybind(byte hotbar, in PointerInputCmdHandler.PointerInputCmdArgs args)
+        {
+            Ui?.HandleChangeHotbarKeybind(hotbar, args);
+        }
+
+        private void OnPlayerDetached(EntityUid uid, ActionsComponent component, PlayerDetachedEvent args)
+        {
+            if (Ui == null) return;
+            _uiManager.StateRoot.RemoveChild(Ui);
+            Ui = null;
+        }
+
+        private void OnPlayerAttached(EntityUid uid, ActionsComponent component, PlayerAttachedEvent args)
+        {
+            Assignments = new(Hotbars, Slots);
+            Ui = new ActionsUI(this, component);
+            _uiManager.StateRoot.AddChild(Ui);
+            UIDirty = true;
         }
```

Comment:
this is extremely gross but guess not much can be done until IGameHud is given the boot.

Reply:
Yeah I tried to avoid making UI changes where possible, so this is all old code that will hopefully be improved when the hud/toolbar are redone.